### PR TITLE
feat: add API and tests for Empty Sector Update support

### PR DESCRIPTION
**Pull request by cryptonemo**

### CODE REVIEW

**Review by ZenGround0 — APPROVED** [review]
Didn't review rust parts but the generated go stuff looks right.

**Comment by ZenGround0** [comment]
@cryptonemo do we typically also add the `proofs.go` wrappers in this PR or do we just do there generated artifacts first and then come back for another pass grouping things together for exposure to lotus?

I'm currently planning on making another PR to do this but if you think it should be done here I'll work here directly.

**Comment by cryptonemo** [comment]
Didn't @Kubuxu do it already?

**Comment by ZenGround0** [comment]
I could be missing it since I don't know this area very well.  But it doesn't look like it.  The place where this seems to happen is proofs.go and this PR doesn't add changes to it other than shuffling imports: https://github.com/filecoin-project/filecoin-ffi/pull/208/files#diff-2433240531e3632252181237effb99adf947c648e290a76b81de485ba30cce86R14

**Comment by cryptonemo** [comment]
> I could be missing it since I don't know this area very well. But it doesn't look like it. The place where this seems to happen is proofs.go and this PR doesn't add changes to it other than shuffling imports: https://github.com/filecoin-project/filecoin-ffi/pull/208/files#diff-2433240531e3632252181237effb99adf947c648e290a76b81de485ba30cce86R14

Maybe I'm missing something, but I thought this was what we needed: https://github.com/filecoin-project/filecoin-ffi/blob/empty-sector-update/sector_update.go

**Comment by ZenGround0** [comment]
Nice purely my mistake, we're all good.

**Comment by cryptonemo** [comment]
> Nice purely my mistake, we're all good.

No worries, thanks for sanity checking.

**Comment by ZenGround0** [comment]
@cryptonemo again I might be missing something important but this seems ready for merge from my end.  Are there proofs things holding you up from taking this out of draft?

**Comment by cryptonemo** [comment]
> @cryptonemo again I might be missing something important but this seems ready for merge from my end. Are there proofs things holding you up from taking this out of draft?

Yes, sorry I haven't documented this well.  This PR depends on the following (ordered):

1) Get final performance numbers on https://github.com/filecoin-project/rust-fil-proofs/pull/1528
2) Merge https://github.com/filecoin-project/rust-fil-proofs/pull/1528
3) Move https://github.com/filecoin-project/rust-fil-proofs/pull/1519 out of draft
4) Move https://github.com/filecoin-project/rust-filecoin-proofs-api/pull/64 out of draft
5) Move *this* PR out of draft

Technically the PRs have had review, but I'd really like to confirm the GPU work is ready and get that merged before moving forward (I did the testing in the past, but my hardware started failing after the final revision and I'm working on getting new hardware up to speed currently).  This is not a requirement, but something I'd like to do.  I don't want to delay anything external on this though.

**Comment by ZenGround0** [comment]
Thanks, this is very clear.  I'm not particularly blocked on this from lotus end as working with submodule on a branch is pretty easy.

**Comment by BlocksOnAChain** [comment]
> > @cryptonemo again I might be missing something important but this seems ready for merge from my end. Are there proofs things holding you up from taking this out of the draft?
> 
> Yes, sorry I haven't documented this well. This PR depends on the following (ordered):
> 
> 1. Get final performance numbers on [Empty sector update tree r gpu rust-fil-proofs#1528](https://github.com/filecoin-project/rust-fil-proofs/pull/1528)
> 2. Merge [Empty sector update tree r gpu rust-fil-proofs#1528](https://github.com/filecoin-project/rust-fil-proofs/pull/1528)
> 3. Move [Add support for Empty Sector Update proofs (for SnapDeals) rust-fil-proofs#1519](https://github.com/filecoin-project/rust-fil-proofs/pull/1519) out of draft
> 4. Move [feat: add API calls required for empty sector update proofs rust-filecoin-proofs-api#64](https://github.com/filecoin-project/rust-filecoin-proofs-api/pull/64) out of draft
> 5. Move _this_ PR out of draft
> 
> Technically the PRs have had a review, but I'd really like to confirm the GPU work is ready and get that merged before moving forward (I did the testing in the past, but my hardware started failing after the final revision and I'm working on getting new hardware up to speed currently). This is not a requirement, but something I'd like to do. I don't want to delay anything external on this though."

@cryptonemo just saw that we made a push on all the things listed as a dependent, should we also follow up with moving this PR out of the draft? 
Or do we still need to "confirm the GPU work" as you mentioned?
FYI: @Kubuxu

**Comment by cryptonemo** [comment]
> @cryptonemo just saw that we made a push on all the things listed as a dependent, should we also follow up with moving this PR out of the draft? Or do we still need to "confirm the GPU work" as you mentioned? FYI: @Kubuxu

@BlocksOnAChain This PR was moved out of draft last week.  I believe it's good to go on the review front.

**Comment by BlocksOnAChain** [comment]
@cryptonemo got it, will follow up on the review so we can try to close it out in the next couple of days, if possible.

**Comment by cryptonemo** [comment]
Requires: https://github.com/filecoin-project/rust-fil-proofs/pull/1519 and https://github.com/filecoin-project/rust-filecoin-proofs-api/pull/64

File: rust/Cargo.toml
Comment (DrPeterVanNostrand) [inline rust/Cargo.toml:62]: Remember to un-patch once related PR's are merged.

**Comment by arajasek** [comment]
Lotus should update to the latest commit in this PR when it lands

File: rust/src/proofs/api.rs
Comment (vmx) [inline rust/src/proofs/api.rs:1418]: I find it better not to put doc comments that contain only a TODO. This way tools can report if docs are missing..

**Review by vmx — COMMENTED** [review]
On the FFI I always fear that we leak memory. Though it looks like you follow the pattern of other calls, and if they don't leak, probably the new ones also won't leak.

I guess we should really have some leak check on CI to be sure.

**Pull request merged**
